query: allow replica to provide arbitrary continue position

denesb opened the pull request:
Currently, we use the last row in the query result set as the position where the query is continued from on the next page. Since only live rows make it into query result set, this mandates the query to be stopped on a live row on the replica, lest any dead rows or tombstones processed after the live rows, would have to be re-processed on the next page (and the saved reader would have to be thrown away due to position mismatch). This requirement of having to stop on a live row is problematic with datasets which have lots of dead rows or tombstones, especially if these form a prefix. In the extreme case, a query can time out before it can process a single live row and the data-set becomes effectively unreadable until compaction gets rid of the tombstones.
This series prepares the way for the solution: it allows the replica to determine what position the query should continue from on the next page. This position can be that of a dead row, if the query stopped on a dead row. For now, the replica supplies the same position that would have been obtained with looking at the last row in the result set, this series merely introduces the infrastructure for transferring a position together with the query result, and it prepares the paging logic to make use of this position. If the coordinator is not prepared for the new field, it will simply fall-back to the old way of looking at the last row in the result set. As I said for now this is still the same as the content of the new field so there is no problem in mixed clusters.

Refs: https://github.com/scylladb/scylla/issues/3672
Refs: https://github.com/scylladb/scylla/issues/7689

/cc @bhalevy @tgrabiec

denesb commented:
@nyh or @psarna please review my alternator changes.

psarna commented on alternator/serialization.cc:
```diff
@@ -161,32 +161,37 @@ bytes get_key_column_value(const rjson::value& item, const column_definition& co
-// entry, whose key is the type (expected to match the key column's type)
-// and the value is the encoded value.
-bytes get_key_from_typed_value(const rjson::value& key_typed_value, const column_definition& column) {
+static const rjson::value& get_typed_value(const rjson::value& key_typed_value, std::string type_str, std::string_view name) {
```
nitpick: I know that the previous code is based on std::string (historical reasons, probably), but perhaps we could just use string_view here? In this function you don't really need to claim ownership of type_str for any reason.

denesb replied:
I think I used `std::string_view` initially but had problems with `rjson`. I will try this again.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/923/

psarna commented on alternator/executor.cc:
```diff
@@ -3573,6 +3573,11 @@ static rjson::value encode_paging_state(const schema& schema, const service::pag
             ++exploded_ck_it;
         }
     }
+    //FIXME conflicts with column names
```
I consider this fixme a blocker for alternator.

As a first step, we should validate that it works on a subset of drivers (at least boto3 via our pytest suite), i.e. to make sure that the drivers don't do any validation, since you're going to return some column values as if they were keys, while they are not defined as such in the schema.


And then, if that works and we agree it's an acceptable divergence, we need to make sure there aren't any conflicts. Probably by reserving some magic column names unlikely to be ever used by anyone, in spirit of `:scylla:paging:region` and `:scylla:paging:weight`, unless somebody has a better suggestion.

Meanwhile, **if** it's trivial to skip alternator in this series and move it to a follow-up, perhaps it's the way to go, in case we're blocked on this discussion for a while. Perhaps the alternator code path could temporarily opt-out from getting a position "between" live rows, and just push for the first live one or eof?

avikivity replied:
It's better to resolve this now IMO. This is a delicate issue and it's better to get it right rather than kick it down the road.

denesb replied:
I thought it would be a blocker. I don't think we can kick this down the road. Although the new feature is not used in this patchset yet and alternator could just opt out, later it would cause extra complexity (explicit flag to opt out) and alternator would keep suffering from workloads that other API will handle well.

I ran all the alternator tests with this: is that enough of a validation for the driver or is there some extra tests I should run?

psarna replied:
@nyh? I think that it's moderately safe to assume that the boto3 driver is Amazon's reference implementation, and it already contains lots of driver-side validation, so if the magic was accepted there, it's safe to treat this divergence as safe in compatibility terms. Personally I think that all drivers would treat this information similarly to how we manage our opaque paging state in CQL - just blindly copy it from the result into the next paged request, without any deep validation or changing anything.

Naturally, we also need to replace `region` and `weight` with custom, reserved names for the attributes, preferably ones containing reserved characters. Ref: https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.NamingRulesDataTypes.html

denesb replied:
There is *some* validation, because my version had this JSON (example):
```json
{
    ...
    "weight": 1,
    "region": "clustered",
    ...
}
```
and the alternator driver did not like that one bit. I had to adjust these new fields to have the same value-object form that the other fields have. The driver doesn't seem to check the content though, only the form, i.e. the tests passed happily after I sorted out this issue and made sure I interpreted keys without weight/region as exclusive keys (like the "exclusiveSortKey" says).

psarna replied:
Technically, at least according to https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.NamingRulesDataTypes.html:
```
Although DynamoDB allows you to use these reserved words and special characters for names, we recommend that you avoid doing so (...)
```
given that, in order to be 100% correct, perhaps we should validate early and check each attribute name to ensure that it isn't on our private reserved list, or, e.g., that it doesn't start with `:scylla:`, which we consider a reserved prefix. Otherwise it will be possible to craft a table which breaks this paging implementation. On the other hand, it's extremely unlikely that somebody would ever create a partition key named `:scylla:paging:weight`, so perhaps we can just let this one slip and not introduce needless checks on the fast path...

psarna replied:
(I'll unresolve this conversation for a short while until I get a response for my latest comment, feel free to reresolve it later)

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/922/

avikivity commented on alternator/executor.cc:
```diff
@@ -3573,6 +3573,11 @@ static rjson::value encode_paging_state(const schema& schema, const service::pag
+    rjson::add(last_evaluated_key, "region", rjson::empty_object());
+    rjson::add(last_evaluated_key["region"], "S", rjson::from_string(to_string(pos.region())));
+    rjson::add(last_evaluated_key, "weight", rjson::empty_object());
+    rjson::add(last_evaluated_key["weight"], "N", static_cast<int>(pos.get_bound_weight()));
```
Might be solved by using a prefix that is an illegal attribute name (alternator terminology for columns).

avikivity commented on idl/paging_state.idl.hh:
```diff
@@ -18,6 +31,8 @@ class paging_state {
     uint32_t get_remaining_high_bits() [[version 4.3]] = 0;
     uint32_t get_rows_fetched_for_last_partition_high_bits() [[version 4.3]] = 0;
+    bound_weight get_clustering_key_weight() [[version 5.1]] = bound_weight::equal;
+    partition_region get_partition_region() [[version 5.1]] = partition_region::partition_start;
```
Isn't 'clustered' more consistent with the current state as a default? So if we read paging_state from an old coordinator we get the old behavior.

avikivity replied:
btw, this will be tricky to test, I guess we'll need a dtest with 1MB rows to have good control over page breaks.

denesb replied:
Yes, `clustered` should be a better default. The dtest suite has good paging tests, I remember when making reads stateful, they caught a lot of bugs.

avikivity commented on service/pager/query_pagers.cc:
```diff
@@ -141,16 +142,19 @@ future<result<service::storage_proxy::coordinator_query_result>> query_pager::do
+            if (_last_pos.has_key()) {
+                next_pos = reversed ? position_in_partition_view::before_key(_last_pos) : position_in_partition_view::after_key(_last_pos);
+            }
+            query::trim_clustering_row_ranges_to(*_schema, row_ranges, next_pos, reversed);
```
You adjust the bound, but only if has_key(). Shouldn't  an adjustment be done regardless of the partition region? For example, if we were in the static row region, we should advance to the clustering region (before all rows).

denesb replied:
The static row is always included in the query result, even if we read just the last row in the partition.
This adjusted position is only used to adjust the clustering row ranges in the slice, positions outside of the clustering domain has no impact on this.

avikivity commented on full_position.hh:
```diff
@@ -0,0 +1,40 @@
+struct full_position;
+
+struct full_position_view {
+    const partition_key& partition;
```
view?

denesb replied:
I will attempt to use it although I seem to recall that `partition_key_view` is not a first class citizen when it comes to comparing keys for example. We don't have good views for keys.

avikivity commented on idl/query.idl.hh:
```diff
@@ -27,6 +27,7 @@ class qr_partition stub [[writable]] {
 
 class query_result stub [[writable]] {
     utils::chunked_vector<qr_partition> partitions; // in ring order
+    std::optional<full_position> last_position [[version 5.1]];
```
What happens on the last page? Do we serialize it anyway? The last page is very common (for short queries that fit in a single page).

denesb replied:
We try to avoid it. In the single partition query code-path we only serialize this if there is a live querier object that is going to be saved and hence we believe there is more data. I'm not sure I added this condition to the range scan code-path, I have to check.

avikivity commented:
Looks good.

nyh commented:
I need to review this code (and the discussion), but so far I only had a thought that it may be relevant to Alternator's #7933.

kostja commented:
* Botond Dénes ***@***.***> [22/06/17 16:34]:
> Currently, we use the last row in the query result set as the position where the query is continued from on the next page. Since only live rows make it into query result set, this mandates the query to be stopped on a live row on the replica, lest any dead rows or tombstones processed after the live rows, would have to be re-processed on the next page (and the saved reader would have to be thrown away due to position mismatch). This requirement of having to stop on a live row is problematic with datasets which have lots of dead rows or tombstones, especially if these form a prefix. In the extreme case, a query can time out before it can process a single live row and the data-set becomes effectively unreadable until compaction gets rid of the tombstones.
> This series prepares the way for the solution: it allows the replica to determine what position the query should continue from on the next page. This position can be that of a dead row, if the query stopped on a dead row. For now, the replica supplies the same position that would have been obtained with looking at the last row in the result set, this series merely introduces the infrastructure for transferring a position together with the query result, and it prepares the paging logic to make use of this position. If the coordinator is not prepared for the new field, it will simply fall-back to the old way of looking at the last row in the result set. As I said for now this is still the same as the content of the new field so there is no problem in mixed clusters.

I am just curious what is the purpose of this from the client
point of view? While this makes paging response latency more
predictable, it increases the overall amount of work done to fetch
the result set, since the number of round trips between the client
and the server is going to be higher.

I still think that the only solution going forward for the dead
tombstones is to prune them more aggressively in tables which have
lots of dead rows. If we're concerned
with data resurrection, we could do it along with an on-demand
repair that runs side by side or precedes compaction.  Keeping the
garbage and teaching all parts of the system deal with it is an
architectural dead end - we're implementing features to deal with
a problem which competition doesn't have in the first place.

> Refs: https://github.com/scylladb/scylla/issues/3672
> Refs: https://github.com/scylladb/scylla/issues/7689
> 
> /cc @bhalevy @tgrabiec 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/scylladb/scylla/pull/10829
> 
> -- Commit Summary --
> 
>   * mutation_fragment.hh: move operator&lt;&lt;(partition_region) to position_in_partition.hh
>   * position_in_partition: add to_string(partition_region) and parse_partition_region()
>   * service/pagers/query_pagers.cc: fix indentation
>   * alternator/serialization: extract value object parsing logic
>   * service/paging: use position_in_partition instead of clustering_key for last row
>   * idl: move position_in_partition into own header
>   * introduce full_position
>   * compaction_mutation: add state accessor
>   * compact_mutation: s/_last_clustering_pos/_last_pos/
>   * querier: rely on compactor for position tracking
>   * querier: use full_position instead of ad-hoc struct
>   * multishard_mutation_query: defer creating result builder until needed
>   * idl/query: add last_position to query_result
>   * mutlishard_mutation_query: propagate compaction state to result builder
>   * mutation_compactor: add current_full_position() convenience accessor
>   * query: have replica provide the last position
> 
> -- File Changes --
> 
>     M alternator/executor.cc (17)
>     M alternator/serialization.cc (56)
>     M alternator/serialization.hh (3)
>     M configure.py (1)
>     M cql3/statements/select_statement.cc (6)
>     A full_position.hh (40)
>     M idl/paging_state.idl.hh (15)
>     M idl/partition_checksum.idl.hh (19)
>     A idl/position_in_partition.idl.hh (31)
>     M idl/query.idl.hh (1)
>     M message/messaging_service.cc (3)
>     M multishard_mutation_query.cc (67)
>     M mutation_compactor.hh (37)
>     M mutation_fragment.cc (29)
>     M mutation_fragment.hh (1)
>     M mutation_partition.cc (6)
>     M position_in_partition.hh (4)
>     M querier.cc (23)
>     M querier.hh (89)
>     M query-result-reader.hh (18)
>     M query-result-writer.hh (14)
>     M query.cc (15)
>     M repair/repair.cc (1)
>     M repair/row_level.cc (1)
>     M replica/table.cc (7)
>     M service/pager/paging_state.cc (20)
>     M service/pager/paging_state.hh (30)
>     M service/pager/query_pager.hh (2)
>     M service/pager/query_pagers.cc (437)
>     M test/boost/secondary_index_test.cc (6)
> 
> -- Patch Links --
> 
> https://github.com/scylladb/scylla/pull/10829.patch
> https://github.com/scylladb/scylla/pull/10829.diff
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/scylladb/scylla/pull/10829
> You are receiving this because you are subscribed to this thread.
> 
> Message ID: ***@***.***&gt;

-- 
Konstantin Osipov, Moscow, Russia

bhalevy commented:
> I still think that the only solution going forward for the dead tombstones is to prune them more aggressively in tables which have lots of dead rows.
> If we're concerned with data resurrection, we could do it along with an on-demand repair that runs side by side or precedes compaction. 

What would be the threshold?  You don't want to compact too aggressively, especially if repair is required (to prevent data resurrection), otherwise this will cause write amplification and waste cpu and memory cycles.

And suppose you're running this compaction and get a query while at it, can you ensure there will be no timeouts due to long prefixes of tombstones?  You could be in a read-amplification scenario in which the partition is spread over many sstables, taking a long time to incorporate all reads together (yes, compaction could be more aggressive to prevent that too).

The bottom line is that we need the ability to return pages with no live rows before the query times out, for keep-alive reasons (that is true also for scans), and at the same time we need to continue to improve and prevent that, efficiently.

kostja commented:
@bhalevy We need to add to our storage format a different tombstone marker than there is today, a tombstone marker that would mean "look no further". When we see such a tombstone marker it means that we can discard any older tombstone for that record. Like we already do with individual rows after Tomek's patch int he cache, we should track rows which have too much garbage in them, and agressively repair them, and replace with a fully compacted row + such tombstone marker. That would allow even standard regular compaction get rid of the excessive garbage quickly, since compaction will know that for these records there is no need to preserve all of the older tombstones.

denesb commented:
This is already possible and works: it is called major compaction. But it is quite heavy-weight.

bhalevy commented:
> @bhalevy We need to add to our storage format a different tombstone marker than there is today, a tombstone marker that would mean "look no further". When we see such a tombstone marker it means that we can discard any older tombstone for that record. Like we already do with individual rows after Tomek's patch int he cache, we should track rows which have too much garbage in them, and agressively repair them, and replace with a fully compacted row + such tombstone marker. That would allow even standard regular compaction get rid of the excessive garbage quickly, since compaction will know that for these records there is no need to preserve all of the older tombstones.

We already purge tombstones as much as we can and safe.
One man's garbage is another man's treasure, you know :)
I.e. we need to keep a tombstone in one sstable if it may shadow data in another sstable.
If we know it doesn't shadow any data in any other sstable, and repair already ran for this token since the tombstone was written, we simply toss it away.  When we end up with many range tombstones or dead rows is that we can't merge them for various reasons.  As Botond said, we may need to do cross-tier compaction (in the form of major compaction, or ICS SAG in the 2 largest tiers), or we must wait for repair, and repair could have had a problem.  Previously, we needed to wait for gc_grace_seconds, but this is nor a factor any more since we require positive indication that repair ran successfully.
Bottom line is that there's no magic potion here that we're missing.

With compaction groups and tablets I think that tuning compaction per table / tablet will allow us to do a better job at compacting tombstones with data as the scope will be narrower (given smaller token ranges per compaction group).

Also, see my proposal for [Incremental Repair and Compaction](https://docs.google.com/document/d/1d82H4BCNR-rEm6YYEt4mPcsLMR9i0eMLWe7eEKlbO3I/edit#heading=h.5m9cvehprxxe) that will better synchronize the two, exactly for the purpose of getting rid of tombstones as fast as possible, and with the lowest write amplification overhead.
Another problem that's tackled there is getting rid of expired data efficiently using sort of expiry-time windows in the fully-repaired tier, so that we can get rid of expired data in a single pass (rather than keeping it as tombstones).

avikivity commented:
> I still think that the only solution going forward for the dead tombstones is to prune them more aggressively in tables which have lots of dead rows. If we're concerned with data resurrection, we could do it along with an on-demand repair that runs side by side or precedes compaction.

Consider this workload:

```
for i in range(10000000):
   DELETE FROM tab WHERE id = :i

SELECT * FROM tab
```

There is no way we can schedule a compaction in time to prevent the SELECT from timing out.

Having tombstone-aware compaction/repair is a good idea (compaction is tombstone-aware to some extend), but they are background processes with no guarantee on timeliness. So we must have something in the foreground that can be certain to work.

kostja commented:
@avi for your example one would expect a range tombstone. I understand that range tombstones can't span multiple partitions today, but can we merge adjacent partitions into a range on a single shard?

kostja commented:
@avi it's going to be easier to reason about things if we stop thinking about tombstone timestamps as timestamps, but assume they are just special log sequence numbers. So a range tombstone can not be coalesced with an adjacent range tombstone with a different LSN only if we may expect that rows at lower LSM layers may have a higher LSN, which is not possible unless we deal with custom timestamps or unbound time drift. The goal of an ongoing row-based on demain repair is to make these two factors bounded - make sure that all nodes time is syncrhonized and that the data doesn't have any timestamps from the future.

kostja commented:
I'd begin by extending the sstable format with more data:
- row timestmamp. this would be max(timestamp of each cell). 
- partition timestamp, same, max(timestamp of each row)
- sstable min/max timestamp

The goal is to see if we can expire data not waiting for repair. And the other kind of tombstone marker, one that purges data regardless of repair, could be used as a means to this end.

denesb commented:
@kostja we have to wait for repair, regardless of what new data we add to our on-file format. We have to keep tombstones around until we are sure all other replicas have seen said tombstone. We used to wait gc grace period and assume this was enough for this, now we wait until repair.

denesb commented:
v2:
* alternator: use string_view in `get_typed_value()`
* alternator: more elaborate name with reserved characters for the new fields
* default region to clustering in idl/paging_state.hh
* Use `partition_key_view` in `full_position_view`
* Only serialize last position in range-scans if we have more data

avikivity commented:
> @avi it's going to be easier to reason about things if we stop thinking about tombstone timestamps as timestamps, but assume they are just special log sequence numbers. So a range tombstone can not be coalesced with an adjacent range tombstone with a different LSN only if we may expect that rows at lower LSM layers may have a higher LSN, which is not possible unless we deal with custom timestamps or unbound time drift. The goal of an ongoing row-based on demain repair is to make these two factors bounded - make sure that all nodes time is syncrhonized and that the data doesn't have any timestamps from the future.

It's not possible to do this with eventual consistency. Tombstones can arrive with out-of-order timestamps.

Even if timestamps are in-order, it's not possible to convert discrete tombstones to range tombstone for some types (and some workloads, e.g. `DELETE FROM tab WHERE id = :i * 10`

kostja commented:
> @kostja we have to wait for repair, regardless of what new data we add to our on-file format. We have to keep tombstones around until we are sure all other replicas have seen said tombstone. We used to wait gc grace period and assume this was enough for this, now we wait until repair.

Neither is necessary if you know your writes reach all replicas. Neither is necessary if you know you have made a read-repair of this specific row. Does it begin to make sense now?

denesb commented:
> > @kostja we have to wait for repair, regardless of what new data we add to our on-file format. We have to keep tombstones around until we are sure all other replicas have seen said tombstone. We used to wait gc grace period and assume this was enough for this, now we wait until repair.
> 
> Neither is necessary if you know your writes reach all replicas. Neither is necessary if you know you have made a read-repair of this specific row. Does it begin to make sense now?

This sounds like a CL=ALL write with read repair enabled, i.e. we contact all replicas owning this row. We allow users to tune CL, will we have to drop this. We also allow them to set custom timestamp which is another rats-nest of problems.
What I'm saying is this: what you propose could work and could work well: but would be slower than current eventual consistent writes. I don't know how much slower and whether this would require a new write mode, or at least a new class of CL. We plan to support raft which solves a lot of our problems but we still want to support the old eventual consistency model for those users who prefer latency over everything else. I don't know if what you propose would be something like this, or if we could just seamlessly integrate it into the current write path, without sacrificing much performance.

nyh commented:
> I need to review this code (and the discussion), but so far I only had a thought that it may be relevant to Alternator's #7933.

I would like to ask that you please clarify the two "Refs" in the commit message. Does it fix those issues or not, and if so, why not? If not, can we say if fixes **some** issue, perhaps a new issue that you'll create just to say what you are fixing?

To be more concrete, issue #7933 which I mentioned above is about a specific Alternator use case where you have two rows with a million rows between them, and "Scan" starting in the first row, and want to see that the first page of the scan to contain only one row, not two (because if it contains two it means the scan must have processed all million rows). Can we write a similar test case for CQL? Will this patch fix it? If so, you can open an issue about that use case, and have this patch Fixes it.

bhalevy commented:
> it is good for rows which have lots of garbage, so that we can compact them without compacting the entire table.

The issue is with partitions, not rows.
Even if read-repair did a full partition compaction on the fly and wrote the compacted partition to disk, we don't have a way to poke holes in the existing sstables that hold data from this partition, so we won't gave to re-read it from them.

When discussing this in the past one lightly-baked idea that floated around was to extend the b-tree from cache to disk, and then updates to a particular token will reach it's respective block and then we can compact those blocks in a log-structured way similar to what we do today but with better locality of access so de-duplication should happen faster and more efficiently (with less write amplification).

denesb commented:
> > I need to review this code (and the discussion), but so far I only had a thought that it may be relevant to Alternator's #7933.
> 
> I would like to ask that you please clarify the two "Refs" in the commit message. Does it fix those issues or not, and if so, why not? If not, can we say if fixes **some** issue, perhaps a new issue that you'll create just to say what you are fixing?
> 
> To be more concrete, issue #7933 which I mentioned above is about a specific Alternator use case where you have two rows with a million rows between them, and "Scan" starting in the first row, and want to see that the first page of the scan to contain only one row, not two (because if it contains two it means the scan must have processed all million rows). Can we write a similar test case for CQL? Will this patch fix it? If so, you can open an issue about that use case, and have this patch Fixes it.

Like I mentioned in the cover-letter, this patch just introduces the infrastructure that makes the fix possible, but doesn't include the fix itself yet. That will be a separate PR.

avikivity commented:
> > @kostja we have to wait for repair, regardless of what new data we add to our on-file format. We have to keep tombstones around until we are sure all other replicas have seen said tombstone. We used to wait gc grace period and assume this was enough for this, now we wait until repair.
> 
> Neither is necessary if you know your writes reach all replicas.

But, you don't know that.

Moreover, until compaction happens, the tombstones have to be kept in order to delete the live data in the other sstables. This is true even for a single-node configuration.


> Neither is necessary if you know you have made a read-repair of this specific row. Does it begin to make sense now?

I understand the motivation, but it just doesn't work with LSM trees. The tombstones are there and it takes a lot of work to remove them.

kostja commented:
* Avi Kivity ***@***.***> [22/06/21 19:02]:
> > Neither is necessary if you know you have made a read-repair of this specific row. Does it begin to make sense now?
> I understand the motivation, but it just doesn't work with LSM trees. The tombstones are there and it takes a lot of work to remove them.

I'd say it doesn't work with coordinator-based replication and
arbitrary timestamps. It works fine with LSM trees.

-- 
Konstantin Osipov, Moscow, Russia

avikivity commented:
> * Avi Kivity ***@***.***> [22/06/21 19:02]:
> > Neither is necessary if you know you have made a read-repair of this specific row. Does it begin to make sense now? I understand the motivation, but it just doesn't work with LSM trees. The tombstones are there and it takes a lot of work to remove them.
> I'd say it doesn't work with coordinator-based replication and arbitrary timestamps. 

Well, we have coordinator based replication and it will take a lot of work to remove it. Probably more work than in this series and its follow-ups.

> It works fine with LSM trees.

I don't understand how. LSMs accumulate tombstones. They can be removed, but the latency of this removal is much greater than typical query latency.

> […](#)
> -- Konstantin Osipov, Moscow, Russia

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/979/

avikivity commented:
Looks fine, but cover letter needs to detail how mixed clusters were tested for compatibility.

avikivity commented:
> * Avi Kivity ***@***.***> [22/06/21 19:39]:
> I don't understand how. LSMs accumulate tombstones. They can be removed, but the latency of this removal is much greater than typical query latency.
> The latency depends on what LSM levels the tombstones are pruned. Higher levels are compacted more frequently, so the goal is to make sure that most tombstones can be removed without reaching the bottom of the tree. The goal is to collapse/coalesce tombstones without waiting for a major compaction.

It's not possible in general. Consider old data that made it to the highest level (largest files). Then a bunch of tombstones entered the lowest level (smallest files). To get rid of the tombstones you need to merge the largest file with the smallest file. That takes time. In addition you need to make sure all the tombstones were propagated to all nodes, including the nodes that are currently down or unreachable.

> […](#)
> -- Konstantin Osipov, Moscow, Russia

kostja commented:
* Konstantin Osipov ***@***.***> [22/06/21 21:02]:
> > I don't understand how. LSMs accumulate tombstones. They can be
> > removed, but the latency of this removal is much greater than
> > typical query latency.
> 
> The latency depends on what LSM levels the tombstones are pruned.
> Higher levels are compacted more frequently, so the goal is to
> make sure that most tombstones can be removed without reaching the
> bottom of the tree. The goal is to collapse/coalesce tombstones without
> waiting for a major compaction.

Let me provide some back of the envelope calculation. Imagine your
workload is 50% insert and 50% delete and you manage to prune 90%
of tombstones without waiting for them to reach the last level of the LSM tree.
Provided that all interim levels combined are ~1/10th of the last
level, your overall footprint shrunk by 40%+ and write amp
shrunk multifold.

-- 
Konstantin Osipov, Moscow, Russia

kostja commented:
* Avi Kivity ***@***.***> [22/06/21 21:19]:
> > * Avi Kivity ***@***.***> [22/06/21 19:39]:
> > I don't understand how. LSMs accumulate tombstones. They can be removed, but the latency of this removal is much greater than typical query latency.
> > The latency depends on what LSM levels the tombstones are pruned. Higher levels are compacted more frequently, so the goal is to make sure that most tombstones can be removed without reaching the bottom of the tree. The goal is to collapse/coalesce tombstones without waiting for a major compaction.
> 
> It's not possible in general. Consider old data that made it to the highest level (largest files). Then a bunch of tombstones entered the lowest level (smallest files). To get rid of the tombstones you need to merge the largest file with the smallest file. That takes time. In addition you need to make sure all the tombstones were propagated to all nodes, including the nodes that are currently down or unreachable.

You're totally right, it doesn't work in general. But it works for
rows that are frequently overwritten. And most writes have a
normal distribution, meaning the impact of this is significant. 
This is exactly the case you're trying to tackle in this pull
request.

-- 
Konstantin Osipov, Moscow, Russia

nyh reviewed (commented):
I can't make a definitive approve/request-changes on this PR because there is something very fundemental which I don't understand because I'm a little out-of-touch with this area of the code:

A row tombstone has a clustering key, just like a clustering row. So as the last patches explain in their commit message, if we let a query say which was the last clustering key considered - instead of just the last clustering key returned as a live row - then we can resume a query after a tombstone. That I understand. What I don't understand is why beyond that, you needed to switch from a "clustering key" to the concept of "position" - I didn't understand what this switch gives us.

I especially don't understand how this effects Alternator. When resuming an Alternator scan, the start position is explicitly called ExclusiveStartKey - it always should be a clustering key that we should start **after** it (i.t., excluding the given key). So when can we ever use all those weird new options of being before or after or whatever? I don't understand.

kostja commented:
* Avi Kivity ***@***.***> [22/06/21 21:19]:
> > * Avi Kivity ***@***.***> [22/06/21 19:39]:
> > I don't understand how. LSMs accumulate tombstones. They can be removed, but the latency of this removal is much greater than typical query latency.
> > The latency depends on what LSM levels the tombstones are pruned. Higher levels are compacted more frequently, so the goal is to make sure that most tombstones can be removed without reaching the bottom of the tree. The goal is to collapse/coalesce tombstones without waiting for a major compaction.
> 
> It's not possible in general. Consider old data that made it to the highest level (largest files). Then a bunch of tombstones entered the lowest level (smallest files). To get rid of the tombstones you need to merge the largest file with the smallest file. That takes time. In addition you need to make sure all the tombstones were propagated to all nodes, including the nodes that are currently down or unreachable.

The optimization I am talking about, in terms of RocksDB, is
called SingleDelete API:

https://github.com/facebook/rocksdb/wiki/Single-Delete

Here is a mention of it:
http://smalldatum.blogspot.com/2020/01/deletes-are-fast-and-slow-in-lsm.html

There it's used for secondary key maintenance. But the idea has a
more broad application.

In this blog there is some general discussion of DeleteRange,
which Scylla already has:
http://rocksdb.org/blog/2018/11/21/delete-range.html

In Tarantool, a similar technique is employed. The main idea is
that there are two kinds of tombstones: the optimized ones, which
are allowed to be pruned in interim levels in certain
circumstances, and the general purpose ones.

The second big idea is that "compaction" happens not in the
background, but during read or write, when we observe the row has
got too much garbage. In this case we compact *just that row* and
put it into LSM as a new "full" write. This new write is exactly the 
kind of write which can be deleted along with the sigle-delete
tombstone when they meet each other in an interim level.

-- 
Konstantin Osipov, Moscow, Russia

avikivity commented:
Single-Delete looks like a great optimization for local tables that satisfy the no-overwrite requirement (paxos table?). We should implement it. However, it doesn't apply if overwrites are possible, which they are in the general case. It also doesn't work for eventually consistent tables, and doesn't work if the data is not in memtable when the delete happens.

Running compaction for the row range being queried it problematic. First, the tombstone prefix is arbitrarily long, and so the latency of the operation is arbitrarily long. So even if we do it, we still need to return in a timely manner like this series prepares for. Second, suppose you compact a range without compacting all of the sstables. You need to record that the compacted range subsumes all the sstables that participated in the compaction. In order to make use of this information, you need to read sstables sequentially, because any sstable you read might tell you that you don't need to read older ones. That can increase latency significantly.

Cassandra had a similar optimization but removed it in apache/cassandra@e2ecdf268a82fa3ac0f4c9fe77ab35bca33cc72a, because they didn't have this extra information recorded (and so they had to read all sstables regardless).

kostja commented:
* Avi Kivity ***@***.***> [22/06/22 13:00]:
> Single-Delete looks like a great optimization for local tables that satisfy the no-overwrite requirement (paxos table?). We should implement it. However, it doesn't apply if overwrites are possible, which they are in the general case. It also doesn't work for eventually consistent tables, and doesn't work if the data is not in memtable when the delete happens.
> 
> Running compaction for the row range being queried it problematic. First, the tombstone prefix is arbitrarily long, and so the latency of the operation is arbitrarily long. So even if we do it, we still need to return in a timely manner like this series prepares for. Second, suppose you compact a range without compacting all of the sstables. You need to record that the compacted range subsumes all the sstables that participated in the compaction. In order to make use of this information, you need to read sstables sequentially, because any sstable you read might tell you that you don't need to read older ones. That can increase latency significantly.

I think compacting a row range, meaning physically overwriting the
files and creating new ones, is a dead end. Instead, I think, we should just be
overwriting the compacted range by inserting it into the memtable,
like it was a regular put. This doesn't have to happen online,
i.e. on read or write, but when we notice that the range has got
too much garbage in it, we can trigger a background operation of
fetching the old versions from all LSM layers and writing a
compacted representation of all of them.

> Cassandra had a similar optimization but removed it in ***@***.***, because they didn't have this extra information recorded (and so they had to read all sstables regardless).

-- 
Konstantin Osipov, Moscow, Russia

denesb commented:
> On 21 Jun 2022, at 20:40, Avi Kivity ***@***.***> wrote:
> 
> ﻿
> Looks fine, but cover letter needs to detail how mixed clusters were tested for compatibility.
> 

In theory there shouldn’t be any issue. We added fields to two structures: the paging state and query result. In both cases the new fields are optional and the code handles them missing (new node) and to ignore them (old node). I can do a regression test just to be on the safe side.

> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.

avikivity commented:
> * Avi Kivity ***@***.***> [22/06/22 13:00]:
> Single-Delete looks like a great optimization for local tables that satisfy the no-overwrite requirement (paxos table?). We should implement it. However, it doesn't apply if overwrites are possible, which they are in the general case. It also doesn't work for eventually consistent tables, and doesn't work if the data is not in memtable when the delete happens. Running compaction for the row range being queried it problematic. First, the tombstone prefix is arbitrarily long, and so the latency of the operation is arbitrarily long. So even if we do it, we still need to return in a timely manner like this series prepares for. Second, suppose you compact a range without compacting all of the sstables. You need to record that the compacted range subsumes all the sstables that participated in the compaction. In order to make use of this information, you need to read sstables sequentially, because any sstable you read might tell you that you don't need to read older ones. That can increase latency significantly.
> I think compacting a row range, meaning physically overwriting the files and creating new ones, is a dead end. Instead, I think, we should just be overwriting the compacted range by inserting it into the memtable, like it was a regular put. This doesn't have to happen online, i.e. on read or write, but when we notice that the range has got too much garbage in it, we can trigger a background operation of fetching the old versions from all LSM layers and writing a compacted representation of all of them.

Inserting it into the memtable doesn't help, because the memtable isn't enough to satisfy reads. We must read for sstables as well. We could add information that the sstables are subsumed, but then we'd need to persist it when the memtable is flushed and we're back at the original problem.

Note that the cache will receive a merged version of the row, but the cache can be evicted at any time, and in any case, it's too late, the client already timeouted.

Tombstone management solutions are important, but they don't solve the problem this effort is trying to solve.

> […](#)
> Cassandra had a similar optimization but removed it in ***@***.***, because they didn't have this extra information recorded (and so they had to read all sstables regardless).
> -- Konstantin Osipov, Moscow, Russia

denesb commented on alternator/executor.cc:
```diff
@@ -3562,9 +3562,9 @@ static rjson::value encode_paging_state(const schema& schema, const service::pag
-    auto ck = paging_state.get_clustering_key();
-    if (ck) {
-        auto exploded_ck = ck->explode();
+    auto pos = paging_state.get_position_in_partition();
```
Yes, it is necessary. In addition to row tombstones we want to handle range tombstones as well. Even though, with the current in-memory format of range-tombstone-changes, a large span of range tombstones is unlikely, it is still possible and if we set out to solve spawns of tombstones, it has to include all the different kind of tombstones we have.
Range tombstones never have a position of **at** a certain clustering key, they are always **before** or **after**.

denesb commented on alternator/executor.cc:
```diff
@@ -3573,6 +3573,10 @@ static rjson::value encode_paging_state(const schema& schema, const service::pag
         }
     }
+    rjson::add_with_string_name(last_evaluated_key, scylla_paging_region, rjson::empty_object());
+    rjson::add(last_evaluated_key[scylla_paging_region.data()], "S", rjson::from_string(to_string(pos.region())));
```
> Is this ".data()" neccessary

Yes, `rjson` refuses to work with string views and `operator[]` doesn't have a variant similar to `add_with_string_name` which does work with string view (not that I know of anyway).

denesb replied:
> The bigger problem I have with this code is that means these new names ":scylla:paging:region" now cannot be used for a key column. This is like #5009 was about the ":attrs" attribute - but actually worse, because 5009 was about fixing the internal representation of a key attribute called ":attrs" - but here you invent a key attribute that can never be used.

I understood based on the Alternator naming rules @psarna linked in earlier in this thread that `:` is not allowed in Alternator column names at all. So it is impossible for a **legal** column name to be named ":scylla:paging:region". Do we enforce this rule?

denesb replied:
> Is this ".data()" neccessary?
> 
> The bigger problem I have with this code is that means these new names ":scylla:paging:region" now cannot be used for a key column. This is like #5009 was about the ":attrs" attribute - but actually worse, because 5009 was about fixing the internal representation of a key attribute called ":attrs" - but here you invent a key attribute that can never be used.
> 
> Despite this problem I actually think that adding extra attributes to the LastEvaluatedKey is a good idea **if it works** - I also mentioned this idea in #6278 which laments the fact that Alternator has just page positions and not the full cookie like in CQL. I worried that the DynamoDB driver and/or applications might get confused by extra keys in LastEvalutatedKey and either complain or just forget to pass it on to the next request, but it seemed you tried this now and it works. In any case, we'll need to document this difference somewhere - we can be bitten by it down the line.

Yes, I printed the raw JSON before after parsing it and the new fields were present. The driver does seem to do *some* validation, as it enforced the new fields to have the object-value format that the columns have.

psarna replied:
> > The bigger problem I have with this code is that means these new names ":scylla:paging:region" now cannot be used for a key column. This is like #5009 was about the ":attrs" attribute - but actually worse, because 5009 was about fixing the internal representation of a key attribute called ":attrs" - but here you invent a key attribute that can never be used.
> 
> I understood based on the Alternator naming rules @psarna linked in earlier in this thread that `:` is not allowed in Alternator column names at all. So it is impossible for a **legal** column name to be named ":scylla:paging:region". Do we enforce this rule?

It is allowed according to the link, it's just discouraged, because such names cannot be easily used with bind markers, which start with `:` in DynamoDB. Here's the exact quote (which I already quoted somewhere above, but let me emphasize the important bit:

"""
Although DynamoDB **allows you to use these reserved words and special characters for names**, we recommend that you avoid doing so because you have to define placeholder variables whenever you use these names in an expression. For more information, see [Expression Attribute Names in DynamoDB](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Expressions.ExpressionAttributeNames.html).
"""

denesb replied:
So, where should I add a note that field names starting with the `:scylla:` prefix are reserved?

psarna replied:
I missed the response to this thread, and it looks like this series got merged meanwhile. It means we have a slight regression in alternator, because certain column names have a conflict with the new paging mechanism. I created a post-mortem issue here: https://github.com/scylladb/scylla/issues/10903

nyh replied:
This irony is that I think eventually we may need to add such extra fake key columns (see https://github.com/scylladb/scylla/issues/6278) but I never understood why we needed them **in this pull request**, and I still don't understand. More about this in 10903, maybe one of you can finally explain it to me :-)

denesb commented on alternator/serialization.hh:
```diff
@@ -33,6 +35,9 @@ struct type_representation {
     data_type dtype;
 };
 
+constexpr std::string_view scylla_paging_region(":scylla:paging:region");
```
It is used in `executor.cc` and `serialization.cc`, as the former does the serialization and the latter the deserialization.

denesb commented on idl/paging_state.idl.hh:
```diff
@@ -6,6 +6,19 @@ enum class read_repair_decision : uint8_t {
+    static_row,
+    clustered,
+    partition_end,
+};
```
This is just to support a full `position_in_partition` to be stored in the paging state, rather than just a clustering key. `position_in_partition` is the standard way to represent a position inside a partition. We probably don't need the entire range of positions it can represent, but I think it is beneficial to use the standard type, instead of coming up a narrower ad-hoc one and add conversions on the edges.

denesb commented on alternator/serialization.cc:
```diff
@@ -242,6 +243,36 @@ clustering_key ck_from_json(const rjson::value& item, schema_ptr schema) {
+    const auto region_item = rjson::find(item, scylla_paging_region);
+    const auto weight_item = rjson::find(item, scylla_paging_weight);
+    if (bool(region_item) != bool(weight_item)) {
+        throw api_error::validation("Malformed value object: region and weight has to be either both missing or both present");
```
This is the same phrase that is used by existing code to refer to the object containing a single entry, keyed by the type.

denesb commented:
New in v3:
* Add comment to `get_typed_value()`
* Add value_name argument to `get_typed_value()` so it is not hard-coded to `key column`
* Make the field constants inline too.

I completely lost track of the discussion in this PR due to lack of threading. I hope I addressed all the comments.
I started working on the test on a mixed cluster, but it is not done yet.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1017/

denesb commented:
I did an upgrade test. I wrote a data set with:
```
./scylla-bench -mode=write -workload=sequential -replication-factor=3 -nodes 127.0.0.1,127.0.0.2,127.0.0.3 -clustering-row-count=10000 -clustering-row-size=8096 -partition-count=1000
```
This creates large, 80MB partitions, which should fit many pages if read in full. Then I started a read workload:
```
./scylla-bench -mode=read -workload=uniform -replication-factor=3 -nodes 127.0.0.1,127.0.0.2,127.0.0.3 -clustering-row-count=10000 -duration=10m -rows-per-request=9000 -page-size=100
```
I confirmed that paging is happening as expected, then upgraded the nodes one-by-one to this PR. I observed no read errors or any other errors in the logs.

bhalevy commented:
@avikivity please re-review

avikivity commented:
Please add the test details to the cover letter. The motivation is that if an upgrade problem is eventually traced to this series, we can try to understand what the tests were missing and apply that to similar changes in the future.

denesb commented:
> Please add the test details to the cover letter. The motivation is that if an upgrade problem is eventually traced to this series, we can try to understand what the tests were missing and apply that to similar changes in the future.

Updated cover-letter.

nyh commented:
> I can't make a definitive approve/request-changes on this PR because there is something very fundemental which I don't understand because I'm a little out-of-touch with this area of the code:
> 
> A row tombstone has a clustering key, just like a clustering row. So as the last patches explain in their commit message, if we let a query say which was the last clustering key considered - instead of just the last clustering key returned as a live row - then we can resume a query after a tombstone. That I understand. What I don't understand is why beyond that, you needed to switch from a "clustering key" to the concept of "position" - I didn't understand what this switch gives us.
> 
> I especially don't understand how this effects Alternator. When resuming an Alternator scan, the start position is explicitly called ExclusiveStartKey - it always should be a clustering key that we should start **after** it (i.t., excluding the given key). So when can we ever use all those weird new options of being before or after or whatever? I don't understand.

@denesb this is my comment about Alternator that I guess you missed in all the chatter on this PR (and the reason why I didn't "approve" this PR). I guess I should have been more explicit about Alternator not having static columns or range tombstones.
